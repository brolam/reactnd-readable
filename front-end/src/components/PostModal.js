import React from 'react'
import PropTypes from 'prop-types'

function PostModal(props) {
  const { post, categories } = props
  return (
    <div id="postModal" className="modal modal-open" >
      <div className="modal-dialog">
        <div className="modal-heard modal-post">
          <span onClick={props.onClickBackButton} />
          <input
            type="text"
            placeholder="Title post"
            defaultValue={post.title}
          />
        </div>
        <div className="modal-content modal-post">
          <textarea
            placeholder="Body post"
            defaultValue={post.body} />
          <div className="modal-footer">
            {isNewPost(post) ?
              getFooterToNewPost(categories, props.onSave)
              :
              getFooterToEditPost(post,  props.onSave)
            }
          </div>
        </div>
      </div>
    </div>
  )
}

function isNewPost(post) {
  return post.category ? false : true;
}

function getFooterToNewPost(categories, onSave) {
  return (
    <select onChange={onSave} >
      <option value="none">Save as?</option>
      {categories.map(category => (
        <option key={category.path} value={category.path} >{category.name.toUpperCase() + category.name.slice(1)}</option>
      ))}
    </select>
  )
}

function getFooterToEditPost(post, onSave) {
  return (
    <button className={"save-button " + post.category} href="/"
      onClick={onSave}>Save
    </button>
  )
}

PostModal.propTypes = {
  categories: PropTypes.array.isRequired,
  onClickBackButton: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired
}

export default PostModal