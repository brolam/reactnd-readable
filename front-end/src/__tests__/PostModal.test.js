import React from 'react'
import ReactDOM from 'react-dom'
import PostModal from '../components/PostModal'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

const categories = global.dataForTest.categories
const postUdacity = global.dataForTest.posts[0]

test('render without error', () => {
  ReactDOM.render(
    <PostModal
      post={postUdacity}
      categories={categories}
      onClickBackButton={e => { }}
      onSave={e => { }}
    />, document.createElement('div'));
})

test('last Snapshot to new Post', () => {
  const postModal = renderer.create(
    <PostModal
      post={{}}
      categories={categories}
      onClickBackButton={e => { }}
      onSave={e => { }}
    />);
  let tree = postModal.toJSON();
  expect(tree).toMatchSnapshot();
})

test('last Snapshot to edit Post', () => {
  const postModal = renderer.create(
    <PostModal
      post={postUdacity}
      categories={categories}
      onClickBackButton={e => { }}
      onSave={e => { }}
    />);
  let tree = postModal.toJSON();
  expect(tree).toMatchSnapshot();
})

describe("New Post to valid and to save", () => {
  let postModal
  let inputTitle
  let { isClikedBackButton, didTheSaveEvent, areTheFieldsValidated } = false

  beforeEach(() => {
    postModal = mount(
      <PostModal
        post={{}}
        onClickBackButton={e => { isClikedBackButton = true }}
        categories={categories}
        onSave={fieldsWasValidated => {
          didTheSaveEvent = true
          areTheFieldsValidated = fieldsWasValidated
        }}
      />);
    isClikedBackButton = false;
    didTheSaveEvent = false;
    areTheFieldsValidated = false;
    inputTitle = postModal.find('div [className="modal-heard modal-post"] input')
  });

  test('input title is focusing', () => {
    const elementFocusing = document.activeElement
    expect(inputTitle.instance() === elementFocusing).toBe(true);
  })

  test('title is required', () => {
    inputTitle.instance().value = ' '
    postModal.find('select').simulate('change');
    expect(areTheFieldsValidated).toBe(false);
  })

  test('title must be longer than 10 characters', () => {
    inputTitle.instance().value = '123456789'
    postModal.find('select').simulate('change');
    expect(areTheFieldsValidated, ).toEqual(false);
  })

  test('title must be less than 80 characters', () => {
    inputTitle.instance().value = '1'.repeat(81)
    postModal.find('select').simulate('change');
    expect(areTheFieldsValidated).toEqual(false);
  })

  test('title must be valid', () => {
    inputTitle.instance().value = '1'.repeat(80)
    postModal.find('select').simulate('change');
    expect(areTheFieldsValidated).toEqual(true);
  })

  test('on save event to new Post', () => {
    inputTitle.instance().value = 'Udacity is the best place to learn technology.'
    postModal.find('select').simulate('change');
    expect(didTheSaveEvent).toBe(true);
    expect(areTheFieldsValidated).toBe(true);
  })

  test('on click back button', () => {
    postModal.find('div [className="modal-heard modal-post"] span').simulate('click')
    expect(isClikedBackButton).toBe(true);
  })
})

test('on save event to Edit Post', () => {
  let didTheSaveEvent = false;
  const postModal = mount(
    <PostModal
      post={postUdacity}
      categories={categories}
      onClickBackButton={e => { }}
      onSave={e => { didTheSaveEvent = true }}
    />);
  postModal.find('button [className="save-button udacity"]').simulate('click');
  expect(didTheSaveEvent).toBe(true);
})
