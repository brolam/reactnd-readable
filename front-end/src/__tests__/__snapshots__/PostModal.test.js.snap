// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last Snapshot to edit Post 1`] = `
<div
  className="modal modal-open"
  id="postModal"
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-heard modal-post"
    >
      <span
        onClick={[Function]}
      />
      <input
        autoFocus={true}
        defaultValue="Udacity is the best place to learn technology."
        placeholder="Title post"
        type="text"
      />
    </div>
    <div
      className="modal-content modal-post"
    >
      <textarea
        defaultValue="Everyone says so after all."
        placeholder="Body post"
      />
      <div
        className="modal-footer"
      >
        <button
          className="save-button udacity"
          href="/"
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`last Snapshot to new Post 1`] = `
<div
  className="modal modal-open"
  id="postModal"
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-heard modal-post"
    >
      <span
        onClick={[Function]}
      />
      <input
        autoFocus={true}
        defaultValue={undefined}
        placeholder="Title post"
        type="text"
      />
    </div>
    <div
      className="modal-content modal-post"
    >
      <textarea
        defaultValue={undefined}
        placeholder="Body post"
      />
      <div
        className="modal-footer"
      >
        <select
          onChange={[Function]}
        >
          <option
            value="none"
          >
            Save as?
          </option>
          <option
            value="udacity"
          >
            Udacity
          </option>
          <option
            value="react"
          >
            React
          </option>
          <option
            value="redux"
          >
            Redux
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
